lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -4499,6 +4569,203 @@ void MySQL_HostGroups_Manager::read_only_action(char *hostname, int port, int re
+
+		is_writer = !writer_map.empty();
+
+		if (read_only == 0) {
+			if (is_writer == false) {
+				// the server has read_only=0 (writer), but we can't find any writer, 
+				// so we copy all reader nodes to writer
+				host_server_mapping->copy_if_not_exists(HostGroup_Server_Mapping::Type::WRITER, HostGroup_Server_Mapping::Type::READER);
+
+				if (mysql_thread___monitor_writer_is_also_reader) {
+					// server is also a reader, we copy all nodes from writer to reader (previous reader nodes will be reused)
+					host_server_mapping->copy_if_not_exists(HostGroup_Server_Mapping::Type::READER, HostGroup_Server_Mapping::Type::WRITER);
+				} else {
+					// server can only be a writer
+					host_server_mapping->clear(HostGroup_Server_Mapping::Type::READER);
+				}
+
+				update_mysql_servers_table = true;
+			} else {
+				bool act = false;
+
+				// if the server was RO=0 on the previous check then no action is needed
+				if (host_server_mapping->get_readonly_flag() != 0) {
+					// it is the first time that we detect RO on this server
+					const std::vector<HostGroup_Server_Mapping::Node>& reader_map = host_server_mapping->get(HostGroup_Server_Mapping::Type::READER);
+
+					for (const auto& reader_node : reader_map) {
+						for (const auto& writer_node : writer_map) {
+
+							if (reader_node.writer_hostgroup_id == writer_node.writer_hostgroup_id) {
+								goto __writer_found;
+							}
+						}
+						act = true;
+						break;
+					__writer_found:
+						continue;
+					}
+
+					if (act == false) {
+						// no action required, therefore we set readonly_flag to 0
+						proxy_info("read_only_action_v2() detected RO=0 on server %s:%d for the first time after commit(), but no need to reconfigure\n", hostname.c_str(), port);
+						host_server_mapping->set_readonly_flag(0);
+					}
+				} else {
+					// the server was already detected as RO=0
+					// no action required
+				}
+
+				if (act == true) {	// there are servers either missing, or with stats=OFFLINE_HARD
+
+					// copy all reader nodes to writer
+					host_server_mapping->copy_if_not_exists(HostGroup_Server_Mapping::Type::WRITER, HostGroup_Server_Mapping::Type::READER);
+
+					if (mysql_thread___monitor_writer_is_also_reader) {
+						// server is also a reader, we copy all nodes from writer to reader (previous reader nodes will be reused)
+						host_server_mapping->copy_if_not_exists(HostGroup_Server_Mapping::Type::READER, HostGroup_Server_Mapping::Type::WRITER);
+					} else {
+						// server can only be a writer
+						host_server_mapping->clear(HostGroup_Server_Mapping::Type::READER);
+					}
+
+					update_mysql_servers_table = true;
+				}
+			}
+		} else if (read_only == 1) {
+			if (is_writer) {
+				// the server has read_only=1 (reader), but we find it as writer, so we copy all writer nodes to reader (previous reader nodes will be reused)
```

Comment:
Also following the same logic of the previous comments:

```
proxy_info("Server '%s:%d' found with 'read_only=1' but not found as reader\n", hostname.c_str(), port);
```

Reply:
Done